File: python/evm_state/rpc.py
"""Capture proofs before backprocessing, while recent state is still available."""
import json
import os
import urllib.request

from .proof import VerificationError, address, quantity, unhex, verify_account, keccak256
from .header import encode_rpc_header, verify_header


class RPC:
    def __init__(self, url=None, key=None):
        self.url = url or os.environ.get("RPC_URL", "https://bsc.rpc.pinax.network")
        self.key = key if key is not None else os.environ.get("RPC_API_KEY", os.environ.get("SUBSTREAMS_API_KEY", ""))

    def call(self, method, params):
        headers = {"Content-Type": "application/json"}
        if self.key:
            headers["X-Api-Key"] = self.key
        request = urllib.request.Request(self.url, headers=headers,
            data=json.dumps({"jsonrpc": "2.0", "id": 1, "method": method, "params": params}).encode())
        with urllib.request.urlopen(request, timeout=60) as response:
            result = json.load(response)
        if "error" in result or result.get("result") is None:
            raise VerificationError(f"RPC {method} failed: {result.get('error', 'null result')}")
        return result["result"]

    def capture(self, accounts, block="finalized", expected_hash=None):
        accounts = sorted({address(a) for a in accounts})
        if not accounts:
            raise ValueError("at least one account is required")
        finalized = self.call("eth_getBlockByNumber", ["finalized", False])
        header = finalized if block == "finalized" else self.call("eth_getBlockByNumber", [hex(int(block)), False])
        number = quantity(header["number"], 64)
        if number > quantity(finalized["number"], 64):
            raise VerificationError("checkpoint block is not finalized")
        if expected_hash is not None and unhex(expected_hash, 32) != unhex(header["hash"], 32):
            raise VerificationError("RPC header differs from the expected block hash")
        bundle = {
            "format_version": 1, "chain_id": quantity(self.call("eth_chainId", [])),
            "header": {"number": number, "hash": header["hash"].lower(),
                "parent_hash": header["parentHash"].lower(), "state_root": header["stateRoot"].lower(),
                "timestamp": quantity(header["timestamp"], 64)},
            "header_trust": "operator-pinned-hash" if expected_hash else "provider-finalized-header",
            "accounts": {},
        }
        bundle["header_rlp"] = encode_rpc_header(header)
        verify_header(bundle["header_rlp"], bundle["header"], expected_hash)
        for account in accounts:
            proof = self.call("eth_getProof", [account, [], hex(number)])
            proven = verify_account(header["stateRoot"], account, proof)
            code = self.call("eth_getCode", [account, hex(number)])
            if keccak256(unhex(code)) != proven.code_hash:
                raise VerificationError("RPC bytecode differs from the proven code hash")
            bundle["accounts"][account] = {"proof": proof, "code": code.lower()}
        after = self.call("eth_getBlockByNumber", [hex(number), False])
        if after["hash"].lower() != header["hash"].lower():
            raise VerificationError("checkpoint header changed during proof capture")
        return bundle

File: crates/evm-state/src/rpc.rs
//! Capture proofs against an immutable finalized block before replay begins.
use crate::{
    header::{encode_rpc_header, verify_header},
    proof::{fixed, quantity, string, unhex, verify_account},
};
use alloy_primitives::keccak256;
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use std::{collections::BTreeSet, env, time::Duration};

pub trait RpcCall {
    fn call(&self, method: &str, params: Value) -> Result<Value>;
}

pub struct Rpc {
    url: String,
    key: String,
    http: reqwest::blocking::Client,
}
impl Rpc {
    pub fn new(url: Option<&str>, key: Option<&str>) -> Result<Self> {
        Ok(Self {
            url: url.map(str::to_owned).unwrap_or_else(|| {
                env::var("RPC_URL").unwrap_or_else(|_| "https://bsc.rpc.pinax.network".into())
            }),
            key: key.map(str::to_owned).unwrap_or_else(|| {
                env::var("RPC_API_KEY")
                    .or_else(|_| env::var("SUBSTREAMS_API_KEY"))
                    .unwrap_or_default()
            }),
            http: reqwest::blocking::Client::builder()
                .timeout(Duration::from_secs(60))
                .redirect(reqwest::redirect::Policy::none())
                .build()?,
        })
    }
}
impl RpcCall for Rpc {
    fn call(&self, method: &str, params: Value) -> Result<Value> {
        let mut request = self
            .http
            .post(&self.url)
            .json(&json!({"jsonrpc":"2.0","id":1,"method":method,"params":params}));
        if !self.key.is_empty() {
            request = request.header("X-Api-Key", &self.key);
        }
        let response = request
            .send()
            .map_err(|_| anyhow::anyhow!("RPC {method} transport failed"))?;
        ensure!(
            response.status().is_success(),
            "RPC {method} failed (HTTP {})",
            response.status().as_u16()
        );
        let result: Value = response.json().context("invalid RPC JSON response")?;
        ensure!(
            result.get("error").is_none() && !result["result"].is_null(),
            "RPC {method} returned an error or null result"
        );
        Ok(result["result"].clone())
    }
}

pub fn capture(
    rpc: &impl RpcCall,
    accounts: impl IntoIterator<Item = impl AsRef<str>>,
    block: Option<u64>,
    expected_hash: Option<&str>,
) -> Result<Value> {
    let accounts = accounts
        .into_iter()
        .map(|a| crate::proof::address(a.as_ref()))
        .collect::<Result<BTreeSet<_>>>()?;
    ensure!(!accounts.is_empty(), "at least one account is required");
    let finalized = rpc.call("eth_getBlockByNumber", json!(["finalized", false]))?;
    let header = if let Some(block) = block {
        rpc.call(
            "eth_getBlockByNumber",
            json!([format!("0x{block:x}"), false]),
        )?
    } else {
        finalized.clone()
    };
    let number = quantity(string(&header, "number")?, 64)?.to::<u64>();
    ensure!(
        number <= quantity(string(&finalized, "number")?, 64)?.to::<u64>(),
        "checkpoint block is not finalized"
    );
    if let Some(hash) = expected_hash {
        ensure!(
            fixed::<32>(hash)? == fixed::<32>(string(&header, "hash")?)?,
            "RPC header differs from the expected block hash"
        );
    }
    let chain = rpc.call("eth_chainId", json!([]))?;
    let chain = quantity(chain.as_str().context("invalid chain ID")?, 256)?;
    let mut bundle = json!({"format_version":1,"chain_id":serde_json::from_str::<Value>(&chain.to_string())?,
        "header":{"number":number,"hash":string(&header,"hash")?.to_ascii_lowercase(),
            "parent_hash":string(&header,"parentHash")?.to_ascii_lowercase(),"state_root":string(&header,"stateRoot")?.to_ascii_lowercase(),
            "timestamp":quantity(string(&header,"timestamp")?,64)?.to::<u64>()},
        "header_trust":if expected_hash.is_some() {"operator-pinned-hash"} else {"provider-finalized-header"},
        "accounts":{},"header_rlp":encode_rpc_header(&header)?});
    verify_header(
        string(&bundle, "header_rlp")?,
        &bundle["header"],
        expected_hash,
    )?;
    for account in accounts {
        let proof = rpc.call(
            "eth_getProof",
            json!([account, [], format!("0x{number:x}")]),
        )?;
        let proven = verify_account(string(&header, "stateRoot")?, &account, &proof)?;
        let code = rpc.call("eth_getCode", json!([account, format!("0x{number:x}")]))?;
        let code = code.as_str().context("invalid RPC bytecode")?;
        ensure!(
            keccak256(unhex(code)?) == proven.code_hash,
            "RPC bytecode differs from the proven code hash"
        );
        bundle["accounts"][account] = json!({"proof":proof,"code":code.to_ascii_lowercase()});
    }
    let after = rpc.call(
        "eth_getBlockByNumber",
        json!([format!("0x{number:x}"), false]),
    )?;
    ensure!(
        string(&after, "hash")?.eq_ignore_ascii_case(string(&header, "hash")?),
        "checkpoint header changed during proof capture"
    );
    Ok(bundle)
}
